File: backend_go_archive/internal/domain/metrics/service.go
package metrics

import (
	"context"
	"fmt"
	"math"
	"time"

	"github.com/admin/turbo-potato/backend/internal/domain/goals"
	"github.com/admin/turbo-potato/backend/internal/domain/rules"
	"github.com/admin/turbo-potato/backend/internal/domain/transactions"
)

type TransactionsReader interface {
	List(ctx context.Context) ([]transactions.Transaction, error)
}

type GoalsReader interface {
	List(ctx context.Context) ([]goals.Goal, error)
}

type RulesReader interface {
	ListFixedCostRules(ctx context.Context) ([]rules.FixedCostRule, error)
}

type Clock interface {
	Now() time.Time
}

type Service struct {
	transactions TransactionsReader
	goals        GoalsReader
	rules        RulesReader
	clock        Clock
}

func NewService(transactions TransactionsReader, goals GoalsReader, rules RulesReader, clock Clock) *Service {
	return &Service{
		transactions: transactions,
		goals:        goals,
		rules:        rules,
		clock:        clock,
	}
}

func (s *Service) Summary(ctx context.Context) (Summary, error) {
	allTransactions, err := s.transactions.List(ctx)
	if err != nil {
		return Summary{}, err
	}

	allGoals, err := s.goals.List(ctx)
	if err != nil {
		return Summary{}, err
	}

	allRules, err := s.rules.ListFixedCostRules(ctx)
	if err != nil {
		return Summary{}, err
	}

	now := s.clock.Now().UTC()
	sts := calculateSTS(allTransactions, now)
	anomaly := calculateAnomaly(allTransactions, now)
	goalPace, goalVelocity, goalETA := calculateGoalPace(allTransactions, allGoals, now)
	fixedCostLoad, runwayMonths := calculateOperatingMetrics(allTransactions, allRules, now)

	return Summary{
		STS:      sts,
		Anomaly:  anomaly,
		GoalPace: goalPace,
		OperatingPosture: OperatingPosture{
			Status: postureStatus(sts.Progress, fixedCostLoad),
			Items: []SummaryItem{
				{Label: "Runway", Value: fmt.Sprintf("%.1f months", runwayMonths)},
				{Label: "Fixed-cost load", Value: fmt.Sprintf("%d%%", fixedCostLoad)},
				{Label: "Goal velocity", Value: goalVelocity},
				{Label: "ETA", Value: goalETA},
			},
		},
		Baselines: buildBaselines(allTransactions, allGoals, allRules, now),
	}, nil
}

func calculateSTS(items []transactions.Transaction, now time.Time) MetricValue {
	const monthlyVariableBudget = 9300000.0

	currentMonthSpent := 0.0
	for _, item := range items {
		if item.Status == transactions.StatusReverted || item.Type != transactions.TypeExpense || item.IsFixed {
			continue
		}
		if item.OccurredAt.Year() == now.Year() && item.OccurredAt.Month() == now.Month() {
			currentMonthSpent += float64(item.Amount)
		}
	}

	daysInMonth := time.Date(now.Year(), now.Month()+1, 0, 0, 0, 0, 0, time.UTC).Day()
	remainingDays := max(1, daysInMonth-now.Day()+1)
	stsValue := (monthlyVariableBudget - currentMonthSpent) / float64(remainingDays)
	if stsValue < 0 {
		stsValue = 0
	}

	progress := int(math.Round((stsValue / (monthlyVariableBudget / float64(daysInMonth))) * 100))
	progress = min(100, max(0, progress))

	return MetricValue{
		Label:    "STS Today",
		Value:    compactCurrency(int64(stsValue)),
		Caption:  "Daily spend allowance based on remaining monthly variable budget.",
		Progress: progress,
		Status:   statusFromProgress(progress),
	}
}

func calculateAnomaly(items []transactions.Transaction, now time.Time) MetricValue {
	todaySpend := 0.0
	history := make([]float64, 0)

	for _, item := range items {
		if item.Status == transactions.StatusReverted || item.Type != transactions.TypeExpense || item.IsFixed {
			continue
		}

		if sameDay(item.OccurredAt, now) {
			todaySpend += float64(item.Amount)
		}

		if item.OccurredAt.Before(now) && item.OccurredAt.After(now.AddDate(0, 0, -30)) {
			history = append(history, float64(item.Amount))
		}
	}

	mean := average(history)
	stddev := standardDeviation(history, mean)
	zScore := 0.0
	if stddev > 0 {
		zScore = (todaySpend - mean) / stddev
	}

	progress := min(100, max(0, int(math.Round((math.Abs(zScore)/2.5)*100))))
	return MetricValue{
		Label:    "Anomaly Score",
		Value:    fmt.Sprintf("%.2f", zScore),
		Caption:  "Variable-spend pacing versus your recent baseline.",
		Progress: progress,
		Status:   anomalyStatus(zScore),
	}
}

func calculateGoalPace(items []transactions.Transaction, goalsList []goals.Goal, now time.Time) (MetricValue, string, string) {
	if len(goalsList) == 0 {
		return MetricValue{
			Label:    "Goal Pace",
			Value:    "0%",
			Caption:  "No active goals yet.",
			Progress: 0,
			Status:   "idle",
		}, "0/mo", "N/A"
	}

	activeGoal := goalsList[0]
	totalTransferred := int64(0)
	recentTransferred := int64(0)

	for _, item := range items {
		if item.Status == transactions.StatusReverted || item.Type != transactions.TypeTransfer {
			continue
		}
		if item.GoalName != activeGoal.Name {
			continue
		}

		totalTransferred += item.Amount
		if item.OccurredAt.After(now.AddDate(0, -3, 0)) {
			recentTransferred += item.Amount
		}
	}

	progress := 0
	if activeGoal.TargetAmount > 0 {
		progress = min(100, max(0, int(math.Round((float64(totalTransferred)/float64(activeGoal.TargetAmount))*100))))
	}

	monthlyVelocity := float64(recentTransferred) / 3.0
	eta := "N/A"
	if monthlyVelocity > 0 && activeGoal.TargetAmount > totalTransferred {
		monthsRemaining := math.Ceil(float64(activeGoal.TargetAmount-totalTransferred) / monthlyVelocity)
		eta = now.AddDate(0, int(monthsRemaining), 0).Format("Jan 2006")
	}

	return MetricValue{
			Label:    "Goal Pace",
			Value:    fmt.Sprintf("%d%%", progress),
			Caption:  "Current savings velocity relative to target timeline.",
			Progress: progress,
			Status:   statusFromProgress(progress),
		},
		compactCurrency(int64(monthlyVelocity)) + "/mo",
		eta
}

func calculateOperatingMetrics(items []transactions.Transaction, rulesList []rules.FixedCostRule, now time.Time) (int, float64) {
	fixedSpent := 0.0
	income := 0.0
	liquid := 0.0

	for _, item := range items {
		if item.Status == transactions.StatusReverted {
			continue
		}

		if item.Type == transactions.TypeExpense && item.IsFixed && item.OccurredAt.Year() == now.Year() && item.OccurredAt.Month() == now.Month() {
			fixedSpent += float64(item.Amount)
		}

		if item.Type == transactions.TypeIncome && item.OccurredAt.After(now.AddDate(0, -3, 0)) {
			income += float64(item.Amount)
			liquid += float64(item.Amount) * 0.3
		}
	}

	if fixedSpent == 0 {
		for _, rule := range rulesList {
			if rule.IsActive {
				fixedSpent += float64(rule.ExpectedAmount)
			}
		}
	}

	avgMonthlyIncome := income / 3
	fixedCostLoad := 0
	if avgMonthlyIncome > 0 {
		fixedCostLoad = int(math.Round((fixedSpent / avgMonthlyIncome) * 100))
	}

	runwayMonths := 0.0
	if fixedSpent > 0 {
		runwayMonths = liquid / fixedSpent
	}

	return fixedCostLoad, runwayMonths
}

func buildBaselines(items []transactions.Transaction, goalsList []goals.Goal, rulesList []rules.FixedCostRule, now time.Time) []BaselineSeries {
	variableSpendSeries := make([]int, 0, 12)
	fixedCostSeries := make([]int, 0, 12)
	goalVelocitySeries := make([]int, 0, 12)

	for monthOffset := 11; monthOffset >= 0; monthOffset-- {
		monthTime := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC).AddDate(0, -monthOffset, 0)
		variableSpendSeries = append(variableSpendSeries, int(monthlyVariableSpend(items, monthTime)))
		fixedCostSeries = append(fixedCostSeries, int(monthlyFixedSpend(items, rulesList, monthTime, now)))
		goalVelocitySeries = append(goalVelocitySeries, int(monthlyGoalTransfers(items, goalsList, monthTime)))
	}

	return []BaselineSeries{
		{
			Label:       "Variable spend",
			Description: "Rolling monthly pace for variable expenses.",
			Values:      normalizeSeries(variableSpendSeries),
			Current:     compactCurrency(int64(variableSpendSeries[len(variableSpendSeries)-1])),
			Delta:       seriesDelta(variableSpendSeries),
			ColorToken:  "var(--tp-danger)",
		},
		{
			Label:       "Fixed-cost load",
			Description: "Monthly fixed obligations against current operating structure.",
			Values:      normalizeSeries(fixedCostSeries),
			Current:     compactCurrency(int64(fixedCostSeries[len(fixedCostSeries)-1])),
			Delta:       seriesDelta(fixedCostSeries),
			ColorToken:  "var(--tp-accent)",
		},
		{
			Label:       "Goal velocity",
			Description: "Monthly transfer momentum toward active goals.",
			Values:      normalizeSeries(goalVelocitySeries),
			Current:     compactCurrency(int64(goalVelocitySeries[len(goalVelocitySeries)-1])),
			Delta:       seriesDelta(goalVelocitySeries),
			ColorToken:  "var(--tp-danger-soft)",
		},
	}
}

func monthlyVariableSpend(items []transactions.Transaction, month time.Time) float64 {
	total := 0.0
	for _, item := range items {
		if item.Status == transactions.StatusReverted || item.Type != transactions.TypeExpense || item.IsFixed {
			continue
		}
		if item.OccurredAt.Year() == month.Year() && item.OccurredAt.Month() == month.Month() {
			total += float64(item.Amount)
		}
	}
	return total
}

func monthlyFixedSpend(items []transactions.Transaction, rulesList []rules.FixedCostRule, month time.Time, now time.Time) float64 {
	total := 0.0
	for _, item := range items {
		if item.Status == transactions.StatusReverted || item.Type != transactions.TypeExpense || !item.IsFixed {
			continue
		}
		if item.OccurredAt.Year() == month.Year() && item.OccurredAt.Month() == month.Month() {
			total += float64(item.Amount)
		}
	}
	if total == 0 && month.Year() == now.Year() && month.Month() == now.Month() {
		for _, rule := range rulesList {
			if rule.IsActive {
				total += float64(rule.ExpectedAmount)
			}
		}
	}
	return total
}

func monthlyGoalTransfers(items []transactions.Transaction, goalsList []goals.Goal, month time.Time) float64 {
	if len(goalsList) == 0 {
		return 0
	}
	total := 0.0
	for _, item := range items {
		if item.Status == transactions.StatusReverted || item.Type != transactions.TypeTransfer {
			continue
		}
		if item.OccurredAt.Year() == month.Year() && item.OccurredAt.Month() == month.Month() {
			total += float64(item.Amount)
		}
	}
	return total
}

func normalizeSeries(values []int) []int {
	if len(values) == 0 {
		return values
	}
	minValue, maxValue := values[0], values[0]
	for _, value := range values[1:] {
		if value < minValue {
			minValue = value
		}
		if value > maxValue {
			maxValue = value
		}
	}
	if maxValue == minValue {
		out := make([]int, len(values))
		for i := range out {
			out[i] = 50
		}
		return out
	}
	out := make([]int, len(values))
	for i, value := range values {
		out[i] = int(math.Round((float64(value-minValue) / float64(maxValue-minValue)) * 100))
	}
	return out
}

func seriesDelta(values []int) string {
	if len(values) < 2 {
		return "0% vs previous"
	}
	current := values[len(values)-1]
	previous := values[len(values)-2]
	if previous == 0 {
		return "0% vs previous"
	}
	delta := ((float64(current) - float64(previous)) / float64(previous)) * 100
	return fmt.Sprintf("%+.0f%% vs previous", delta)
}

func compactCurrency(value int64) string {
	if value >= 1000000 {
		return fmt.Sprintf("%.1fm", float64(value)/1000000)
	}
	if value >= 1000 {
		return fmt.Sprintf("%.0fk", float64(value)/1000)
	}
	return fmt.Sprintf("%d", value)
}

func statusFromProgress(progress int) string {
	switch {
	case progress >= 80:
		return "healthy"
	case progress >= 50:
		return "warning"
	default:
		return "critical"
	}
}

func anomalyStatus(z float64) string {
	switch {
	case math.Abs(z) > 1.96:
		return "critical"
	case math.Abs(z) > 1.0:
		return "warning"
	default:
		return "healthy"
	}
}

func postureStatus(stsProgress, fixedCostLoad int) string {
	switch {
	case stsProgress < 40 || fixedCostLoad > 60:
		return "High alert"
	case stsProgress < 70 || fixedCostLoad > 40:
		return "Moderate risk"
	default:
		return "Stable"
	}
}

func average(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	total := 0.0
	for _, value := range values {
		total += value
	}
	return total / float64(len(values))
}

func standardDeviation(values []float64, mean float64) float64 {
	if len(values) == 0 {
		return 0
	}
	sum := 0.0
	for _, value := range values {
		diff := value - mean
		sum += diff * diff
	}
	return math.Sqrt(sum / float64(len(values)))
}

func sameDay(a, b time.Time) bool {
	return a.Year() == b.Year() && a.Month() == b.Month() && a.Day() == b.Day()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: backend/app/domain/metrics/service.py
from __future__ import annotations

import math
from datetime import UTC, datetime, timedelta
from typing import Protocol

from app.core.runtime import Clock
from app.domain.goals.model import Goal
from app.domain.metrics.model import BaselineSeries, MetricValue, OperatingPosture, Summary, SummaryItem
from app.domain.rules.model import FixedCostRule
from app.domain.sources.model import Source
from app.domain.transactions.model import Transaction


class TransactionsReader(Protocol):
    def list(self) -> list[Transaction]: ...


class GoalsReader(Protocol):
    def list(self) -> list[Goal]: ...


class RulesReader(Protocol):
    def list_fixed_cost_rules(self) -> list[FixedCostRule]: ...


class SourcesReader(Protocol):
    def list(self) -> list[Source]: ...


class MetricsService:
    def __init__(
        self,
        transactions: TransactionsReader,
        goals: GoalsReader,
        rules: RulesReader,
        sources: SourcesReader,
        clock: Clock,
    ) -> None:
        self._transactions = transactions
        self._goals = goals
        self._rules = rules
        self._sources = sources
        self._clock = clock

    def summary(self) -> Summary:
        all_transactions = self._transactions.list()
        all_goals = self._goals.list()
        all_rules = self._rules.list_fixed_cost_rules()
        all_sources = self._sources.list()
        now = self._clock.now().astimezone(UTC)

        sts = calculate_sts(all_transactions, all_goals, all_rules, now)
        anomaly = calculate_anomaly(all_transactions, now)
        tar = calculate_tar(all_transactions, now)
        goal_pace, goal_velocity, goal_eta = calculate_goal_pace(all_transactions, all_goals, now)
        fixed_cost_load, runway_months = calculate_operating_metrics(all_transactions, all_rules, all_sources, now)
        return Summary(
            sts=sts,
            anomaly=anomaly,
            tar=tar,
            goalPace=goal_pace,
            operatingPosture=OperatingPosture(
                status=posture_status(sts.progress, fixed_cost_load, runway_months),
                items=[
                    SummaryItem(label="Runway", value=f"{runway_months:.1f} months"),
                    SummaryItem(label="Fixed-cost load", value=f"{fixed_cost_load}%"),
                    SummaryItem(label="Goal velocity", value=goal_velocity),
                    SummaryItem(label="ETA", value=goal_eta),
                ],
            ),
            baselines=build_baselines(all_transactions, all_goals, all_rules, now),
        )


def calculate_sts(items: list[Transaction], goals_list: list[Goal], rules_list: list[FixedCostRule], now: datetime) -> MetricValue:
    income_baseline = baseline_monthly_income(items, now)
    fixed_commitment = month_fixed_commitment(items, rules_list, now)
    goal_commitment = required_goal_commitment(goals_list, items, now)
    current_month_spent = sum(
        item.amount
        for item in items
        if item.status != "reverted"
        and item.type == "OUT"
        and not item.is_fixed
        and item.occurred_at.year == now.year
        and item.occurred_at.month == now.month
    )
    days_in_month = (datetime(now.year, now.month % 12 + 1, 1, tzinfo=UTC) - timedelta(days=1)).day if now.month != 12 else 31
    remaining_days = max(1, days_in_month - now.day + 1)
    flexible_budget = max(0.0, income_baseline - fixed_commitment - goal_commitment)
    sts_value = max(0.0, (flexible_budget - current_month_spent) / remaining_days)
    target_daily_allowance = flexible_budget / days_in_month if days_in_month > 0 else 0.0
    progress = min(100, max(0, round((sts_value / target_daily_allowance) * 100))) if target_daily_allowance > 0 else 0
    return MetricValue(
        label="STS",
        value=compact_currency(int(sts_value)),
        caption="Safe-to-spend per day after fixed commitments and target funding needs.",
        progress=int(progress),
        status=status_from_progress(int(progress)),
    )


def calculate_anomaly(items: list[Transaction], now: datetime) -> MetricValue:
    today_spend = 0.0
    daily_totals: dict[str, float] = {}

    for day_offset in range(60, -1, -1):
        day = (now - timedelta(days=day_offset)).date().isoformat()
        daily_totals[day] = 0.0

    for item in items:
        if item.status == "reverted" or item.type != "OUT" or item.is_fixed:
            continue
        day_key = item.occurred_at.astimezone(UTC).date().isoformat()
        if day_key in daily_totals:
            daily_totals[day_key] += float(item.amount)
        if same_day(item.occurred_at, now):
            today_spend += item.amount

    today_key = now.date().isoformat()
    history = [value for day, value in daily_totals.items() if day != today_key]
    median_value = median(history)
    mad = median([abs(value - median_value) for value in history])
    z_score = 0.0 if mad == 0 else 0.6745 * (today_spend - median_value) / mad
    progress = min(100, max(0, round((abs(z_score) / 3.5) * 100)))
    return MetricValue(
        label="Anomaly",
        value=f"{z_score:.2f}",
        caption="Robust daily-spend anomaly score versus the last 60 days.",
        progress=int(progress),
        status=anomaly_status(z_score),
    )


def calculate_goal_pace(items: list[Transaction], goals_list: list[Goal], now: datetime) -> tuple[MetricValue, str, str]:
    active_goals = [goal for goal in goals_list if goal.status == "active"]
    if not active_goals:
        return MetricValue(label="Goal Pace", value="0%", caption="No active goals yet.", progress=0, status="idle"), "0/mo", "N/A"
    active_goal = sorted(active_goals, key=lambda goal: (goal.target_date or datetime.max.replace(tzinfo=UTC), goal.start_date))[0]
    total_transferred = 0.0
    recent_transferred = 0.0
    for item in items:
        if item.status == "reverted" or item.type != "TRANSFER" or item.goal_name != active_goal.name:
            continue
        total_transferred += item.amount
        if item.occurred_at > now - timedelta(days=90):
            recent_transferred += item.amount
    monthly_velocity = recent_transferred / 3.0
    remaining = max(0.0, active_goal.target_amount - total_transferred)
    pace_ratio = 0.0
    if active_goal.target_date and active_goal.target_date > now and remaining > 0:
        months_remaining = max(1, month_span(now, active_goal.target_date))
        required_monthly = remaining / months_remaining
        pace_ratio = monthly_velocity / required_monthly if required_monthly > 0 else 1.0
    elif active_goal.target_amount > 0:
        pace_ratio = total_transferred / active_goal.target_amount
    progress = min(100, max(0, round(pace_ratio * 100)))
    monthly_velocity = recent_transferred / 3.0
    eta = "N/A"
    if monthly_velocity > 0 and remaining > 0:
        months_remaining = math.ceil(remaining / monthly_velocity)
        eta_dt = now.replace(day=1) + timedelta(days=32 * months_remaining)
        eta = eta_dt.strftime("%b %Y")
    return (
        MetricValue(
            label="Goal Pace",
            value=f"{int(round(pace_ratio * 100))}%",
            caption="Current goal funding pace versus the rate required to hit target.",
            progress=int(progress),
            status=goal_pace_status(pace_ratio),
        ),
        compact_currency(int(monthly_velocity)) + "/mo",
        eta,
    )


def calculate_tar(items: list[Transaction], now: datetime) -> MetricValue:
    month_income = 0
    month_variable_out = 0
    month_fixed_out = 0

    for item in items:
        if item.status == "reverted" or item.occurred_at.year != now.year or item.occurred_at.month != now.month:
            continue
        if item.type == "IN":
            month_income += item.amount
        elif item.type == "OUT":
            if item.is_fixed:
                month_fixed_out += item.amount
            else:
                month_variable_out += item.amount
    if month_income <= 0:
        return MetricValue(
            label="TAR",
            value="0%",
            caption="Net savings rate based on income kept after actual spending.",
            progress=0,
            status="idle",
        )

    retained = month_income - month_variable_out - month_fixed_out
    ratio = retained / month_income
    progress = min(100, max(0, round(ratio * 100)))
    return MetricValue(
        label="TAR",
        value=f"{int(round(ratio * 100))}%",
        caption="Net savings rate after actual consumption; transfers are not double-counted.",
        progress=int(progress),
        status=savings_rate_status(ratio),
    )


def calculate_operating_metrics(
    items: list[Transaction],
    rules_list: list[FixedCostRule],
    sources: list[Source],
    now: datetime,
) -> tuple[int, float]:
    fixed_spent = month_fixed_commitment(items, rules_list, now)
    avg_monthly_income = baseline_monthly_income(items, now)
    liquid = liquid_reserves(sources)
    variable_burn = baseline_monthly_variable_spend(items, now)
    fixed_cost_load = int(round((fixed_spent / avg_monthly_income) * 100)) if avg_monthly_income > 0 else 0
    monthly_burn = fixed_spent + variable_burn
    runway_months = liquid / monthly_burn if monthly_burn > 0 else 0.0
    return fixed_cost_load, runway_months


def build_baselines(items: list[Transaction], goals_list: list[Goal], rules_list: list[FixedCostRule], now: datetime) -> list[BaselineSeries]:
    variable_series: list[int] = []
    fixed_series: list[int] = []
    goal_series: list[int] = []
    for month_offset in range(11, -1, -1):
        month_time = datetime(now.year, now.month, 1, tzinfo=UTC)
        month_shift = month_time.month - month_offset
        year = month_time.year + (month_shift - 1) // 12
        month = ((month_shift - 1) % 12) + 1
        shifted = datetime(year, month, 1, tzinfo=UTC)
        variable_series.append(int(monthly_variable_spend(items, shifted)))
        fixed_series.append(int(monthly_fixed_spend(items, rules_list, shifted, now)))
        goal_series.append(int(monthly_goal_transfers(items, goals_list, shifted)))
    return [
        BaselineSeries(label="Variable spend", description="Rolling monthly pace for variable expenses.", values=normalize_series(variable_series), current=compact_currency(variable_series[-1]), delta=series_delta(variable_series), colorToken="var(--tp-danger)"),
        BaselineSeries(label="Fixed-cost load", description="Monthly fixed obligations against current operating structure.", values=normalize_series(fixed_series), current=compact_currency(fixed_series[-1]), delta=series_delta(fixed_series), colorToken="var(--tp-accent)"),
        BaselineSeries(label="Goal velocity", description="Monthly transfer momentum toward active goals.", values=normalize_series(goal_series), current=compact_currency(goal_series[-1]), delta=series_delta(goal_series), colorToken="var(--tp-danger-soft)"),
    ]


def monthly_variable_spend(items: list[Transaction], month: datetime) -> float:
    return float(sum(item.amount for item in items if item.status != "reverted" and item.type == "OUT" and not item.is_fixed and item.occurred_at.year == month.year and item.occurred_at.month == month.month))


def monthly_fixed_spend(items: list[Transaction], rules_list: list[FixedCostRule], month: datetime, now: datetime) -> float:
    total = float(sum(item.amount for item in items if item.status != "reverted" and item.type == "OUT" and item.is_fixed and item.occurred_at.year == month.year and item.occurred_at.month == month.month))
    if total == 0 and month.year == now.year and month.month == now.month:
        total = float(sum(rule.expected_amount for rule in rules_list if rule.is_active))
    return total


def monthly_goal_transfers(items: list[Transaction], goals_list: list[Goal], month: datetime) -> float:
    if not goals_list:
        return 0.0
    return float(sum(item.amount for item in items if item.status != "reverted" and item.type == "TRANSFER" and item.occurred_at.year == month.year and item.occurred_at.month == month.month))


def normalize_series(values: list[int]) -> list[int]:
    if not values:
        return values
    min_v, max_v = min(values), max(values)
    if max_v == min_v:
        return [50 for _ in values]
    return [round(((value - min_v) / (max_v - min_v)) * 100) for value in values]


def series_delta(values: list[int]) -> str:
    if len(values) < 2 or values[-2] == 0:
        return "0% vs previous"
    delta = ((values[-1] - values[-2]) / values[-2]) * 100
    return f"{delta:+.0f}% vs previous"


def compact_currency(value: int) -> str:
    if value >= 1_000_000:
        return f"{value/1_000_000:.1f}m"
    if value >= 1_000:
        return f"{value/1_000:.0f}k"
    return str(value)


def status_from_progress(progress: int) -> str:
    if progress >= 80:
        return "healthy"
    if progress >= 50:
        return "warning"
    return "critical"


def anomaly_status(z: float) -> str:
    if abs(z) > 1.96:
        return "critical"
    if abs(z) > 1.0:
        return "warning"
    return "healthy"


def posture_status(sts_progress: int, fixed_cost_load: int, runway_months: float) -> str:
    if sts_progress < 45 or fixed_cost_load > 60 or runway_months < 3:
        return "High alert"
    if sts_progress < 70 or fixed_cost_load > 40 or runway_months < 6:
        return "Moderate risk"
    return "Stable"


def average(values: list[float]) -> float:
    return sum(values) / len(values) if values else 0.0


def median(values: list[float]) -> float:
    if not values:
        return 0.0
    ordered = sorted(values)
    middle = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[middle]
    return (ordered[middle - 1] + ordered[middle]) / 2


def same_day(a: datetime, b: datetime) -> bool:
    return a.year == b.year and a.month == b.month and a.day == b.day


def month_start(now: datetime) -> datetime:
    return datetime(now.year, now.month, 1, tzinfo=UTC)


def shift_month(anchor: datetime, offset: int) -> datetime:
    month_index = anchor.month + offset
    year = anchor.year + (month_index - 1) // 12
    month = ((month_index - 1) % 12) + 1
    return datetime(year, month, 1, tzinfo=UTC)


def monthly_income(items: list[Transaction], month: datetime) -> float:
    return float(
        sum(
            item.amount
            for item in items
            if item.status != "reverted" and item.type == "IN" and item.occurred_at.year == month.year and item.occurred_at.month == month.month
        )
    )


def baseline_monthly_income(items: list[Transaction], now: datetime) -> float:
    anchor = month_start(now)
    history = [monthly_income(items, shift_month(anchor, offset)) for offset in (-3, -2, -1)]
    realized = [value for value in history if value > 0]
    if realized:
        return average(realized)
    return monthly_income(items, anchor)


def baseline_monthly_variable_spend(items: list[Transaction], now: datetime) -> float:
    anchor = month_start(now)
    history = [monthly_variable_spend(items, shift_month(anchor, offset)) for offset in (-3, -2, -1)]
    realized = [value for value in history if value > 0]
    if realized:
        return average(realized)
    return monthly_variable_spend(items, anchor)


def month_fixed_commitment(items: list[Transaction], rules_list: list[FixedCostRule], now: datetime) -> float:
    fixed_spent = float(
        sum(
            item.amount
            for item in items
            if item.status != "reverted"
            and item.type == "OUT"
            and item.is_fixed
            and item.occurred_at.year == now.year
            and item.occurred_at.month == now.month
        )
    )
    expected_rules = float(sum(rule.expected_amount for rule in rules_list if rule.is_active))
    return max(fixed_spent, expected_rules)


def total_goal_funded(goal: Goal, items: list[Transaction]) -> float:
    return float(
        sum(
            item.amount
            for item in items
            if item.status != "reverted" and item.type == "TRANSFER" and item.goal_name == goal.name
        )
    )


def required_goal_commitment(goals_list: list[Goal], items: list[Transaction], now: datetime) -> float:
    total = 0.0
    for goal in goals_list:
        if goal.status != "active":
            continue
        funded = total_goal_funded(goal, items)
        remaining = max(0.0, goal.target_amount - funded)
        if remaining <= 0:
            continue
        if goal.target_date and goal.target_date > now:
            months_remaining = max(1, month_span(now, goal.target_date))
            total += remaining / months_remaining
    return total


def month_span(start: datetime, end: datetime) -> int:
    return max(1, (end.year - start.year) * 12 + (end.month - start.month) + (1 if end.day >= start.day else 0))


def goal_pace_status(pace_ratio: float) -> str:
    if pace_ratio >= 1.0:
        return "healthy"
    if pace_ratio >= 0.75:
        return "warning"
    return "critical"


def savings_rate_status(ratio: float) -> str:
    if ratio >= 0.2:
        return "healthy"
    if ratio >= 0.1:
        return "warning"
    return "critical"


def liquid_reserves(sources: list[Source]) -> float:
    reserves = 0.0
    for source in sources:
        if not source.is_active:
            continue
        if source.kind == "gold":
            reserves += source.gold_quantity_chi * source.gold_price_per_chi * 0.85
            continue
        reserves += max(0, source.actual_balance)
    return reserves
